src/lib.rs

```diff
@@ -65,11 +70,30 @@ impl FxHasher {
 
 impl Hasher for FxHasher {
     #[inline]
-    fn write(&mut self, bytes: &[u8]) {
-        for byte in bytes {
-            let i = *byte;
-            self.add_to_hash(i as usize);
+    fn write(&mut self, mut bytes: &[u8]) {
+        #[cfg(target_pointer_width = "32")]
+        let read_usize = |bytes| NativeEndian::read_u32(bytes);
+        #[cfg(target_pointer_width = "64")]
+        let read_usize = |bytes| NativeEndian::read_u64(bytes);
+
+        let mut hash = FxHasher { hash: self.hash };
+        assert!(size_of::<usize>() <= 8);
+        while bytes.len() >= size_of::<usize>() {
+            hash.add_to_hash(read_usize(bytes) as usize);
+            bytes = &bytes[size_of::<usize>()..];
+        }
+        if (size_of::<usize>() > 4) && (bytes.len() >= 4) {
+            hash.add_to_hash(NativeEndian::read_u32(bytes) as usize);
+            bytes = &bytes[4..];
+        }
+        if (size_of::<usize>() > 2) && bytes.len() >= 2 {
+            hash.add_to_hash(NativeEndian::read_u16(bytes) as usize);
+            bytes = &bytes[2..];
+        }
+        if (size_of::<usize>() > 1) && bytes.len() >= 1 {
+            hash.add_to_hash(bytes[0] as usize);
```

Comment:
Doesn't all of this mean that splitting a `write` call changes the hash? IIRC it shouldn't.
Could an `union { bytes: [u8; size_of::<usize>()], usize: usize }` buffer be used instead?

Reply:
That doesn't seem to be a documented nor a useful property.

Reply:
cc @michaelwoerister @Gankro I remember discussions about this property

Note that it's potentially useful to buffer the values if, with e.g. nested enums, you're writing byte-sized values (i.e. discriminants) most of the time, one at a time.

Reply:
Since the `FxHasher` is only used with hash tables, I don't think that the hash must be stable. As long as it is deterministic for our use cases, it's fine, I think. It already treats `(u8, u8)` different from `u16` where a similar argument could be made.

My view is: `FxHasher` should be the absolute fastest for small keys and it should do whatever it can get away with in practice.

Reply:
Yeah, we don't need to do this in this PR. The benchmarks showed that it's an improvement.

As a sidenote, using perf.rlo is a lot more complicated when testing out-of-tree crates...